src/openrct2/world/Map.cpp

```diff
@@ -1539,6 +1545,103 @@ void map_update_tiles()
     }
 }
 
+void mapScheduleCalcHighestTileHeight()
+{
+    gCalcHighestTileHeightScheduled = true;
+}
+
+void mapTryCalcHighestTileHeight(bool fullCheck)
+{
+    if (!fullCheck && !gCalcHighestTileHeightScheduled)
+    {
+        return;
+    }
+
+    int32_t maxTileIndex = MAXIMUM_MAP_SIZE_TECHNICAL * MAXIMUM_MAP_SIZE_TECHNICAL;
+    int32_t tilesToCheck = maxTileIndex;
+
+    // If we're doing a full check of the map then zero out the loop
+    // position to start at tile zero, checking all tiles before returning.
+    // Otherwise check a limited number of tiles per each time the
+    // function is called.
+    if (fullCheck)
+    {
+        gMapHighestTileHeightLoopPosition = 0;
+    }
+    else
+    {
+        tilesToCheck /= 64;
+    }
+
+    for (int32_t i = 0; i < tilesToCheck; i++)
+    {
+        gMapHighestTileHeightLoopPosition++;
+        gMapHighestTileHeightLoopPosition %= maxTileIndex;
+
+        if (gMapHighestTileHeightLoopPosition == 0)
```

Comment:
In order for it to finish it would have to wrap at least once. It would be better to check if ```gMapHighestTileHeightLoopPosition == maxTileIndex```, this loop should only go from 0 to maxTileIndex.
This also makes the use of modulo redundant.